runtime/parser2/parser.go

```diff
@@ -54,7 +54,8 @@ func (p *parser) report(err error) {
 
 func (p *parser) next() {
 	p.pos++
-	p.current = p.tokens[p.pos]
+	p.current = <-p.tokens
```

Comment:
Maybe this should use the two-value result form and check if a value was sent.
If the second boolean result value is false, p.current can be set to `TokenEOF`.
Without it, when the channel was closed, this read will return zero-values instantly.